src/gui/MainApplication.cpp

```diff
@@ -86,8 +86,8 @@ bool MainApplication::event(QEvent* event)
 }
 
 #ifdef LMMS_BUILD_WIN32
-// This can be moved into nativeEventFilter once Qt4 support has been dropped
-bool MainApplication::winEventFilter(MSG* msg, long* result)
+// Should have been a part of nativeEventFilter but keeping here for readability
```

Comment:
Curious about this code comment.  The old comment says:

> This can be moved into nativeEventFilter once Qt4 support has been dropped

Is this statement no longer true?  I feel that the new wording is a bit harder to understand.

> Should have been a part of nativeEventFilter but keeping here for readability

Reply:
https://github.com/LMMS/lmms/pull/7086#pullrequestreview-2041227224
Here

Reply:
I'm only speaking on behalf of the comment, not the code.

Reply:
I can revert the comment changes tho. If I should, leave a 👍 for me to come to a decision.

Reply:
How about a mix of both. 

> This can be moved to nativeEventFilter but kept here for readability

Reply:
Can you help explain to me what the comment means?

Reply:
I am thinking of removing the comment now since you said it's not clear and I am having a hard time coming up with new statements but would be better to have @DomClark 's word (please keep it simple, I can't read through that wall of text.)

Reply:
I've read the comments in #7086.  This is what I'm understanding, please correct me if I'm wrong

* nativeEventFilter is used regardless.
* winEventFilter is windows-only code that was split off for readability

If this is true, we should guard the windows-only stuff, and the comment would mention that it's a helper function for windows (which may simply be self-documenting once the guard is there).

Please let me know if this makes sense or if I'm wrong in these assumptions.

Reply:
Ahh yes, "helper function". That was what I was missing. Thanks for the hint.

Reply:
See the file again. There are guards in the file.

Reply:
My assumptions are wrong, it's all only for Windows.   🤔

Reply:
> See the file again. There are guards in the file.

Yeah I saw this.  I'll wait for Dom to chime in.  I think readability is going to be important here and he seems to understand it best.

Reply:
I disagree that the code "should" be all in one function. All the compiler cares about is whether the code is legal, and all the end user cares about is whether the code is correct (and performant). Both approaches - inlining that function or keeping it separate - satisfy this. Readability and maintainability are what matter for making this choice, and the choice should always be to write readable, maintainable code.

A comment saying "this code is primarily written to be readable" is mostly redundant - why would you not do that, if you had the choice? It only serves as a note that code is safe to refactor in the future, but that should be the default assumption, with comments left when it's not obvious that code needs to be left as-is. I'd just remove the obsolete comment entirely, rather than changing it.

Whether or not keeping it as two functions does indeed improve readability and maintainability is a different question. My preference now (as opposed to whatever it might have been seven years ago when I wrote that code) is to keep them separate. I wrote in https://github.com/LMMS/lmms/pull/7086#pullrequestreview-2041227224:
> Having thought some more about this, I think it makes sense to keep `nativeEventFilter` and `winEventFilter` separate. If we ever need to handle native events for other platforms, we'll likely want to have separate functions for each platform, to prevent `nativeEventFilter` getting unnecessarily complex.

Even while it's only used for a single platform, I still think it makes sense to keep Windows-specific code out of the general event filter.

Reply:
> If we ever need to handle native events for other platforms,

Apologies, I had apparently read https://github.com/LMMS/lmms/pull/7086#pullrequestreview-2041227224 too fast.  To that point, would it make sense to add it with a no-op for other platforms for now, or best left for the future?   Pardon my lack of understanding, I don't really know what this code does.  I assume something with dragging, which might be helpful to have debug messages about when unimplemented.

Reply:
It's to stop VST plugins making the LMMS window transparent. Some plugins (I can't remember which) have a knob to control the opacity of the window containing them, which is useful when they're not embedded and configured to stay on top of the main window. When embedded, it instead results in the main window becoming transparent. This code intercepts any style change messages for the main window and clears the transparency flag before allowing them to be processed further.

I don't see any advantage to adding a no-op event filter for other platforms right now. It won't simplify the code - the Windows-specific bits need to stay there anyway, so we'd just be moving where we put the ifdefs - and can only slow things down (although perhaps imperceptibly).

Reply:
Thanks for the explanation.  Perhaps a small comment, then?  "Prevent VST causing LMMS window to go transparent"?

Reply:
Fixed the comment

Reply:
> Perhaps a small comment, then? "Prevent VST causing LMMS window to go transparent"?

Like the one on line 97?
```cpp
// Prevent plugins making the main window transparent
```

> Fixed the comment

I honestly would just remove it. Don't start with the assumption that there needs to be a comment, and then try to figure out what to put in it. Instead, add a comment when there is something that needs to be said. I wouldn't add a comment to a function to say where it is used - this is prone to becoming out-of-date, and needs updating even when the function itself is not modified. The correct, up-to-date information is easily available from the code itself, either through grep or a built-in IDE tool.

Reply:
> Like the one on line 97?

> ```
> // Prevent plugins making the main window transparent
> ```

Yes, thank you.

I started writing a detailed reply, but I feel an example is clearer.

```cpp
        // win32 implementation 
	bool winEventFilter(MSG* msg, FilterResult* result);
```

```cpp
        // Handle top-level window events.  Note: implementation is platform-specific.
	bool nativeEventFilter(const QByteArray& eventType, void* message, FilterResult* result);
```

... this warrants the helper and helps points to what's happening.  With regards to stating what a function is used for, I have mixed feelings about this, but I don't feel that the discussion would be productive here in this code review, so I've removed it.

Reply:
> `win` is ambiguous, a rename would also suffice

This makes sense. It was named `winEventFilter` originally to override a Qt 4 function of that name, which is no longer a concern.

> `top-level` can be implied, since we're in MainApplication. but `window events` are not as obvious, nor is platform divergence

It's not just window events - it's any native events dispatched on the main thread, of which window events are a part. The handler is attached via `QCoreApplication::installNativeEventFilter`: note that that class isn't part of Qt's widget library, or even GUI library - it's part of the core. See [the documentation](https://doc.qt.io/qt-5/qcoreapplication.html#installNativeEventFilter) for more details.

I would have thought platform divergence would be implied by "native" - the function will be handling native platform events, which of course vary from platform to platform.

Reply:
> I would have thought platform divergence would be implied by "native" - the function will be handling native platform events, which of course vary from platform to platform.

Good point.


> This makes sense. It was named `winEventFilter` originally to override a Qt 4 function of that name, which is no longer a concern.

Right, to handle window events. 😉 

> It's not just window events - it's any native events dispatched on the main thread

I think these comments can do a lot to help.  Instead of trying to remove helpful text, we could add helpful text:

> ```cpp
> // Handle top-level events window events, hotkeys, etc
> ```

... but I do prefer to explain functionality, it provides hansel and gretel breadcrumbs years later.  I digress.